### ISSUE TITLE
Suggestion: Rebase your envvar patch and PR it to main repo.

### ISSUE THREAD

**Issue by IngeniousDox** [Firerat/wine-pba#2]
Andrew just accepted Mike's rebase for staging 3.5. I suggest you rebase your envvar patch on that, and then PR it to the main repo for inclusion. Its a good idea and some of us are happily using it. If Andrew feels you need to change anything, he'll let you know in the PR.

Dox

**Comment by Firerat** [comment]
maybe, but first I want to flesh it out a bit,  i.e. have it check what is given to it
things like ,
geo+cb < vram, 
cb=0, then cb=1 ( cb=0 causes pba to failback to no pba )

**Comment by IngeniousDox** [comment]
Understandable. I noticed with dx9 Starcraft II (dx9), setting stuff high does indeed lower fps. So some sane starting points, and some suggestions about min/max could be good.

Oh, while you are making env var patches, can you add an envvar for the option to enable GL_CLIENT_STORAGE_BIT for NVidia as well? So we can test the difference between using DMA cached memory vs. System Heap memory.

**Comment by Firerat** [comment]
I only have 1024mb vram on my gt620, and just 256mb on my laptop.. so I can't really graph much out

the storage bit is easy
```
diff --git a/dlls/wined3d/buffer_heap.c b/dlls/wined3d/buffer_heap.c
index 9e8f2d799d..c78f114c08 100644
--- a/dlls/wined3d/buffer_heap.c
+++ b/dlls/wined3d/buffer_heap.c
@@ -174,7 +174,7 @@ HRESULT wined3d_buffer_heap_create(struct wined3d_context *context, GLsizeiptr s
     //
     // Thus, we're only going to set this on mesa for now.
     // Hints are awful anyway.
-    if (gl_info->quirks & WINED3D_QUIRK_USE_CLIENT_STORAGE_BIT)
+    if ((gl_info->quirks & WINED3D_QUIRK_USE_CLIENT_STORAGE_BIT) || (getenv("__PBA_FORCE_GL_CLIENT_STORAGE_BIT")))
     {   
         FIXME_(d3d_perf)("PBA: using GL_CLIENT_STORAGE_BIT quirk");
         storage_flags |= GL_CLIENT_STORAGE_BIT;

```
already has a conditional
so, I just added an or `||` which will return true if the getenv "gets anything"
only briefly tested that it doesn't crash,

**Comment by IngeniousDox** [comment]
I wasn't suggesting you would give the values for each game. I figured that if this is main wine-pba, we can start gathering sane defaults, min / maxes for games because people can test it. ;)

Anyways, yes, that was exactly what I was talking about. Just being able to force it on. Although Andrew has the same vid card as me, I would just like to be able to test DMA cached. Since it should take less CPU usage then using system heap...we never got a good chance to test it with current code.

**Comment by Firerat** [comment]
yeah, I know : D
https://github.com/acomminos/wine-pba/issues/47 was my attempt at collecting such data, but low volume. I need to find a more suitable forum.

so how has the storage bit force worked out for you?
I did play with it a little. With the limited testing I did I noticed a small improvement.. but my system has way too many bottlenecks to produce any worthwhile data.

anyway, I've wrapped it up in to the geo/cb heap patch and will do PR soon

**Comment by IngeniousDox** [comment]
For me it using it felt worse. But this was from a very limited test. Friday I'll test in mythic raiding.

**Comment by Firerat** [comment]
ahh yes, that is why I put "for better or for worse" in my commit notes ;)

on another note, are you tagged of chasing staging?
staging is a little ambiguous with the csmt toggle switch in winecfg
it was the dllredirect, it is now labeled as depreciated BUT now toggled off turns upstream CSMT off

that is going to catch a lot of people out

**Comment by IngeniousDox** [comment]
It shouldn't be labeled Depreciated if it is going to toggle normal CSMT. Tell the devs.

**Comment by Firerat** [comment]
yeah, already commented in the commit that changed the toggle to upstream's csmt

they havn't understood yet, bless them

**Comment by IngeniousDox** [comment]
Confirming client_storage_bit on is worse for NVidia. It was the reason why I was getting 45 fps in Overwatch. With system heap I get 70+ constantly.

PS: the way you created the check for __PBA_FORCE_GL_CLIENT_STORAGE_BIT with the ||, it will always enable client_storage_bit, as long as the variable is present, even if it is unset. Since empty string isn't false in C. And setting it to "0" also doesn't make it false etc. Didn't think about it before, but noticed it when I tried to set it to 0 in my env.

**Comment by Firerat** [comment]
unset?
no not true
it is true that setting at 0 would still enable it, but `unset __PBA_FORCE_GL_CLIENT_STORAGE_BIT`
will disable it

**Comment by IngeniousDox** [comment]
Guess I shouldn't have said: unset. Just no value at all, so '\0'.
Exported like: `export _PBA_FORCE_GL_CLIENT_STORAGE_BIT`

Had it set in Lutris, so to disable it, I first set it to 0 (like you do with all variables). Then I tried not setting anything. Both failed. So I had a 2nd look at the OR.

**Comment by Firerat** [comment]
yeah,
the `getenv()` dosn't care about what is in the envvar, just that it exists
ok, so something that will behave more like you expect is
```
( atoi(getenv("__PBA_FORCE_GL_CLIENT_STORAGE_BIT")) == '1' )
```
`man atoi` and `man getenv`

I should probably push that to the fleshed out version
